---
title: "Genomic Variants: Introduction to exomeSeq data"
author: Jineta Banerjee
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: paged 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Getting Started

This document aims to guide the hackathon participants, who have limited exposure to Synapse as a platform and clinical/biological data, through exploration and analysis of the available exomeSeq datasets. The available dataset has been carefully curated to include genomic variant (exomeSeq) data from normal and four different tumor types from three different study initiatives into a single platform (a harmonized dataset). 

The scope of this document is to inform participants:

* how to navigate the Synapse environment to access and work with the data files
* display the key features of the genomic variant (exomeSeq) dataset 
* introduce few general analyses steps of the data and their interpretation

At the end of this document, the participants will be able to download the data and understand various intricacies that exist in this given dataset. We encourage participants to visualize the dataset by faceting on different variables to get an accurate understanding of the dataset. 

## Loading Analytical Packages

Lets first load the R packages we need to query, download, and analyze the data that is stored on Synapse.
```{r lib_synapser, eval=TRUE, results='show', message=FALSE, warning=FALSE, include=TRUE}

library(synapser)
library(synapserutils)
library(BiocManager)
library(gProfileR)
library(tidyverse)
library(DT)
library(maftools)
library(here)
library(ggfortify)
library(stringr)
library(reshape2)

## The login credentials have been stored in .synapseConfig file in the home dir. If you dont have this setup, please make the file before trying to login.
synLogin()

#Make color spectrum
vc_cols = RColorBrewer::brewer.pal(n = 100, name = 'Spectral')
```
&nbsp;

## Accessing or downloading data from Synapse

Please look up the SynID of the relevant data table and access it using the __synTableQuery__ command of *synapser*. This file contains genomic variant data that has been collected from NF1 patients by various researchers as a part of various different studies. 

```{r load data, eval=TRUE, results='show', message=FALSE, warning=FALSE}

exome_data <- synTableQuery("SELECT * FROM syn20812188")$asDataFrame()

```
&nbsp;

# Exploring the data files

The accessed file is a dataframe which contains the genomic variant data collected from normal and tumor samples of four different tumor types. The last few columns of the dataframe contain meta-data associated with the samples. All silent mutations (mutations that do not affect the sequence of amino acids in the protein product of the gene) have been removed to facilitate analysis of mutations that may affect protein sequences.

## Structure of the file

Lets first take a brief look at the contents of the accessed file:
```{r explore data, eval=TRUE, results='show', message=FALSE, warning=FALSE, include=TRUE}

# Clean up dataframe

exome_data$sex[exome_data$sex == "male"] <- "Male"
exome_data$sex[exome_data$sex == "female"] <- "Female"
exome_data <- exome_data[,-c(1:2)]

##Inspect the file
print("Contents of the file:")
head(exome_data)

```

The following table contains descriptions of some of the metadata variables included in the data file:

Var | Description
--- | --- 
 _Hugo_Symbol_  |  names of genes according to HUGO database
 _Entrez_gene_Id_  |  Gene ID according to Entrez Database
 _NCBI_Build_ | Reference genome that was used to align the exomeSeq data
 _Chromosome_ | Chromosome number (range 1-22 and X,Y),  Chr M == mitochondrial genome (absent in exomes with NCBI Build == hg19)
 _species_  |  the source of the specimen
 _age_  |  the age of the patient 
 _sex_  |  the sex of the patient
 _tumorType_  |  the the type of tumor, can be one of 7 different diagnoses
 _isCellLine_  |  indicates whether the origin tissue was a cell line or a patient
 _study_  |  the specific initiative/consortia that the study was a part of
 _id_ | Synapse ID of the sample (unique for each sample)
 
For more information regarding the various columns in the dataset please visit here: https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/#vep

&nbsp;

## Metadata 

The metadata in a healthcare study includes descriptions of the specimens, and the details of the studies they originated from. We can explore the metadata by visualizing various aspects of it to get a sense of the data contained within our dataset and put it in context of the disease and patients.

For example, we can visualize the data distribution according to sex.

```{r explore Metadata1, eval=TRUE, results='show', message=FALSE, warning=FALSE, include=TRUE}

#Plot Metadata
exome_data %>% group_by(Tumor_Sample_Barcode) %>%
ggplot(aes(x=sex))+
    geom_bar(aes(x=sex,fill=as.factor(tumorType)),position='dodge')+
    scale_fill_brewer(palette="Spectral")+
    ggtitle('Exome-seq sample distribution')
```

We can also look at the different types of mutations observed for various genes in our dataset according to the tumor type. The different types of variants recognized in our dataset include:

Variant_classification | Description
--- | ---
Nonsense_Mutation | Mutation leading to change of a coding codon to stop codon
Splice_Site | Mutation leading to change in splice site
Missense_Mutation | Mutation resulting in change in amino acid
In_Frame_Del | Deletion of nucleotides divisible by three leading to deletions of amino acids
In_Frame_Ins | Insertion of nucleotides divisible by three leading to insertion of amino acids
Frame_Shift_Ins | Insertions of nucleotides (not divisible by three) such that codons downstream of the insertion are shifted resulting in a malformed protein or nonsense-mediated decay
Frame_Shift_Del | Deletions of nucleotides (not divisible by three) such that codons downstream of the deletion are shifted resulting in a malformed protein or nonsense-mediated decay
Translation_Start_Site | Mutation causing changes in translation start site
Nonstop_Mutation | SNP in stop codon that disrupts the stop codon causing continued translation

More information regarding variants can be found [here](https://uswest.ensembl.org/info/genome/variation/prediction/classification.html#classes)



```{r waterfallplot, eval=TRUE, results='show', message=FALSE, warning=FALSE, include=TRUE, fig.height=8, fig.width=8}

#Parse out anno and data from csv
anno <- exome_data[, c("Tumor_Sample_Barcode", "id", "age", "assay", "diagnosis", 
                     "individualID", "nf1Genotype","nf2Genotype","organ","isCellLine", "sex", 
                     "species", "specimenID", "study", "studyId", "disease","tumorType")]
anno <- unique(anno)

drop_cols <- c("id", "age", "assay", "diagnosis", 
                     "individualID", "nf1Genotype","nf2Genotype","organ","isCellLine", "sex", 
                     "species", "specimenID", "study", "studyId", "disease","tumorType")
maf <- exome_data[, !(names(exome_data) %in% drop_cols)]

mut_vec <- c(
  "Nonsense_Mutation",
  "Splice_Site",
  "Missense_Mutation",
  "In_Frame_Del",
  "In_Frame_Ins",
  "Frame_Shift_Ins",
  "Frame_Shift_Del",
  "Translation_Start_Site",
  "Nonstop_Mutation",
  "IGR" 
)
# Use read.maf to make maf object
maf_exome <- read.maf(maf = maf, clinicalData = anno, vc_nonSyn = mut_vec, verbose = FALSE) #use maftools to generate visualizations

#Make color spectrum
names(vc_cols) = mut_vec


#Oncoplot / Waterfall plot
oncoplot(maf = maf_exome, top = 30,  colors = vc_cols, fontSize = 0.5, fill = FALSE,
         logColBar = TRUE, clinicalFeatures = "tumorType", sortByAnnotation = TRUE, 
         removeNonMutated = FALSE, bgCol = "#FFFFFF")

```

We can zoom in on a few of our favorite genes to explore the different flavors of mutations that genes may carry. 

```{r explore Metadata2, eval=TRUE, results='show', message=FALSE, warning=FALSE, include=TRUE}
#Plot genes with metadata
analyzeMetdataWithGenes<-function(datafile,prefix, gene_a, ...){
  #look at some marker genes using normalized zScore
  ggplot(subset(datafile,Hugo_Symbol%in%c(gene_a, ...)))+
    geom_jitter(aes(x=Hugo_Symbol, y=Variant_Classification, shape=study,col=tumorType), width =0.1, size=2, height = 0.1)+
    ggtitle(paste(prefix,'Variant types in selected genes', sep = '::'))
}
analyzeMetdataWithGenes(exome_data, "Exome-seq", "NF1", "GFAP", "ASTE1")

```

In the above plot we picked three genes, _nf1_, _gfap_, and _aste1_. Mutations in _nf1_ gene have been strongly correlated with the diagnosis and prognosis of Neurofibromatosis 1. The plot above confirms the presence of various types of mutations in the _nf1_ gene in all of our patient samples. The _gfap_ gene is primarily expressed in astrocytes in the central nervous system (cells that show uncontrolled growth in glioma) as well as in Schwann cells in the peripheral nervous system (cells involved in cutaneous neurofibroma). Unlike the various different kinds of mutations noted in _nf1_, we observe the presence of only missense mutations in _gfap_ in a few of our patient samples. _aste1_ gene encodes a protein that may be involved in epithelial growth factor signaling, a pathway that is implicated in uncontrolled cell growth. We observe the presence of frame shift deletions in this gene in some of our samples.

&nbsp;

# Data Analysis

## Dimensionality Reduction

To explore the data and find meaningful differences between samples, we first perform a simplistic principal component analysis, and explore the genes that are top contributors to the first few principal components. 

We tested the hypothesis whether the number of mutations in a set of genes can be used for clustering the data/samples into meaningful groups. For this, we first created a scoring matrix by converting variant data into numerical variables for PCA. 

```{r r making a scoring matrix, eval=TRUE, results='hide', message=FALSE, warning=FALSE}

# Filter select columns for analysis
select_df <- exome_data[, c("Hugo_Symbol", "Entrez_Gene_Id", "Chromosome", "Start_Position", "Variant_Type", 
                            "Transcript_ID", "HGVSp_Short", "BIOTYPE", "SWISSPROT", "RefSeq", "PolyPhen", 
                            "IMPACT", "FILTER", "id", "assay", "diagnosis", "individualID", "isCellLine", 
                            "organ", "sex", "species", "study", "tumorType", "gnomAD_AF", "gnomAD_AFR_AF",
                            "gnomAD_AMR_AF", "gnomAD_ASJ_AF", "gnomAD_EAS_AF", "gnomAD_FIN_AF", "gnomAD_NFE_AF",
                            "gnomAD_OTH_AF", "gnomAD_SAS_AF")]

# Make dummy variables for PCA
select_df <- select_df %>% mutate(isSNP = as.numeric(select_df$Variant_Type == "SNP")) 
select_df <- select_df %>% mutate(isDEL = as.numeric(select_df$Variant_Type == "DEL")) 
select_df <- select_df %>% mutate(isDNP = as.numeric(select_df$Variant_Type == "DNP"))  
select_df <- select_df %>% mutate(isINS = as.numeric(select_df$Variant_Type == "INS")) 
select_df <- select_df %>% mutate(isONP = as.numeric(select_df$Variant_Type == "ONP"))  
select_df <- select_df %>% mutate(isTNP = as.numeric(select_df$Variant_Type == "TNP")) 
select_df <- select_df %>% mutate(totHits = rowSums(select_df[,c(33:38)]))

# Filter out well known common variants to look for rare mutations, and select mutations with annotated "HIGH" impact on proteins
impact_df <- select_df[(select_df$IMPACT == "HIGH" & select_df$FILTER != "common_variant"),]
```

We used the number of mutational events (_hits_) to the genes as a scoring metric for each gene to do our principal component analysis. Since some genomic variants can be quite common and many of them are well tolerated, we filtered out all known common_variants in the genes and selected for the ones with high impact on proteins to strengthen the signal for PCA. 

```{r PCA, eval=TRUE, results='show', message=FALSE, warning=FALSE}


doPcaHitPlots<-function(datafile, env){
  #create matrix
  combined.mat=reshape2::acast(datafile,Hugo_Symbol~id,value.var="totHits", fun.aggregate = sum)
  #then remove missing
  #missing=which(apply(combined.mat,1,function(x) any(is.na(x))))
  #combined.mat=combined.mat[-missing,]
  
  ###PCA 
  env$tab <- autoplot(prcomp(t(combined.mat))) 

  ###Look at genes contributing to PCs
  loads=prcomp(combined.mat)$x
  
  genes1=rownames(combined.mat)[rev(order(loads[,1]))[1:25]]
  #now take those loadings from pc1
  env$tab1 <- ggplot(subset(datafile,Hugo_Symbol%in%genes1[1:5]))+
                    geom_jitter(aes(x=Hugo_Symbol,y=Variant_Type,col=tumorType,shape=study), width = 0.1, size = 1)+
                    ggtitle('Selected genes from PC1')

  genes2=rownames(combined.mat)[rev(order(loads[,2]))[1:25]]
  #now take those loadings from pc2
  env$tab2 <- ggplot(subset(datafile,Hugo_Symbol%in%genes2[1:5]))+
              geom_jitter(aes(x=Hugo_Symbol,y=Variant_Type,col=tumorType,shape=study), width = 0.1, size = 1)+
              ggtitle('Selected genes from PC2')
}


myEnv <- new.env()
doPcaHitPlots(impact_df, myEnv)
print("PCA: excluding common variants (scoring metric : no. of hits)")
myEnv$tab

```

We notice that the PCA seems to identify two different clusters in the data based on this metric, and also identifies few samples as outliers. The panels below show the different types of mutations observed in some of the genes that contribute to the first and second principal components of the analysis.

```{r }
myEnv$tab1
myEnv$tab2

```

However, even with our best efforts to select relevant mutations, the first principal component explains only 13% of the variance in the data. We encourage the participants to explore other variables present in the dataset and combinations thereof to find more significant sources of variance within the data. 

```{r}

sessionInfo()

```
