---
title: "SciDataRelease_Figures"
author: Jineta Banerjee
date: '`r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document:
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: paged 
    #code_folding: true 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r lib_synapser, echo=FALSE, eval=TRUE, results='hide', message=FALSE, warning=FALSE, include=FALSE}

library(synapser)
library(synapserutils)

library(BiocManager)
library(gProfileR)
library(tidyverse)
library(DT)
library(stringr) 
library(maftools)
library(GenVisR)
library(RColorBrewer)
library(wesanderson)

## The login credentials have been stored in .synapseConfig file in the home dir. If you dont have this setup, please make the file before trying to login.
synLogin()

```

### Summary :

The code below makes figures for the Scientific Data Paper showcasing the JHU Biobank data.

```{r makingmaf files from SynTables, echo=FALSE, eval=TRUE, results='hide', message=FALSE, warning=FALSE}

biobank_public_query <- "SELECT * FROM syn20554939 WHERE study = 'JHU NTAP Biobank' AND FILTER <> 'common_variant' "
#biobank_public_query <- "SELECT * FROM syn20554939 WHERE study = 'JHU NTAP Biobank' "
jhu_biobank_table <- synTableQuery(biobank_public_query)$asDataFrame()
jhu_biobank_table <- jhu_biobank_table %>% 
  dplyr::filter(!specimenID %in% c("2-004 pNF Cell Line", "2-004 NF Cell Line"))

#Make MAF from table dataframe
jhu_public_maf <- jhu_biobank_table[, c(3:135)]
jhu_public_clindata <- unique(jhu_biobank_table[,c(3,136:154)])

#Add sampleType to clinical data
jhu_public_clindata$sampleType <- "Tissue"
jhu_public_clindata$sampleType[jhu_public_clindata$isCellLine==TRUE] <- "Cell Line"
#add xenograft as sample type
jhu_public_clindata$sampleType[(!is.na(str_match(jhu_public_clindata$specimenID,"Xenograft"))) ] <- "Xenograft"

#Make clinical data file for normal and tumor data
jhu_public_clindata_normal <- jhu_public_clindata[(jhu_public_clindata$tumorType == "Normal"),]
jhu_public_clindata_tumor <- jhu_public_clindata[(jhu_public_clindata$tumorType != "Normal"),]

# Make MAF object for Maftools
jhu_public_maf_all <- read.maf(jhu_public_maf, clinicalData = jhu_public_clindata)
jhu_normal_file <- subsetMaf(jhu_public_maf_all, tsb = c(jhu_public_clindata_normal$Tumor_Sample_Barcode))
jhu_tumor_file <- subsetMaf(jhu_public_maf_all, tsb = c(jhu_public_clindata_tumor$Tumor_Sample_Barcode))


##### NOTE #####
# If a sample is normal but has a separate VCF file generated by callers other than VARSCAN, then using its barcode as Tumor_Sample_Barcode will lead to the following :
# -Tumor_Sample_Barcode field has the barcode of the normal sample, matched_norm_sample Alleles are replaced by reference allele, Matched Normal Sample barcode says "NORMAL". Other than these nothing changes.
# If a sample is normal, using its barcode as Matched_normal_sample_barcode will lead to the following:
# - Matched_norm_sample_barcode has the sample barcode, Tumor_sample_barcode says "TUMOR" . This leads to maftools not recognising the right sample maf to merge/annotate
# - The + strand of Tumor_sample_allele is replaced by ref allele, but -strand has - strand allele of sample, matched norm sample alleles are the same as tumor sample alleles if the file were generated with barcode in Tumor_sample_barcode field
# All other fields stay exactly same for both cases. So from now on.. for consistency we will use all samples with their barcode in Tumor_sample_barcode field. The annotations associated with the samples will determine whether a sample was normal or tumor.

```
&nbsp;

### Figure.1 

Make a plot of all the NF1 pathogenic variants from the mafs that we are releasing for the sciData paper (sciDataRelease = TRUE)? This will serve as technical validation for the paper. So something that shows the variants (some samples might not have one?), and whether or not the sample is a pdx, cell line, blood, or tumor.

```{r all_sample_oncoplot, echo= FALSE, eval=TRUE, fig.height=8,fig.width=15, results='show', message=FALSE, warning=FALSE}

###### Make color scheme for plots ########

## Make colors for plots
fabcolors = RColorBrewer::brewer.pal(n = 11,name = 'RdGy')
col = RColorBrewer::brewer.pal(n = 10,name = 'PRGn')
allcolors <- c(fabcolors, col)
sampleID <- sort(unique(as.character(unique(jhu_public_clindata$individualID))))
names(allcolors) = c(sampleID)
allcolors <- list(allcolors)

morecolors1 <- wes_palette("Darjeeling1", n=4, type = "discrete")
tumortypes <- sort(unique(as.character(unique(jhu_public_clindata$tumorType))))
names(morecolors1) <- c(tumortypes)
morecolors1 <- list(morecolors1)

morecolors2 <- wes_palette("Moonrise2", n=3, type = "discrete")
sampletype <- sort(unique(as.character(unique(jhu_public_clindata$sampleType))))
names(morecolors2) = c(sampletype)
morecolors2 <- list(morecolors2)

color_list <- c(allcolors, morecolors1, morecolors2)


####### Plots  ###########

## Oncoplot showing all the samples and all genes that are important for cNF and MPNST

require(maftools)

# Make oncoplot for all samples for specific genes (top column bar shows mutational burden of the samples, row bar on the right shows frequency of mutations per gene)
pdf("Scidata_fig_3_allsamples_oncoplot_03172020_nocommonvar.pdf", width=20,height=8)#,height = 800, width = 1800 )
oncoplot(jhu_public_maf_all, 
         genes = c("NF1","TP53", "CDKN2A", "EED", "SUZ12", 
                   "EZH2", "CHD4", "EPC1", "ASTE1", "DAXX", "ATRX", "PCLO", 
                  "MTAP", "PREX2", "NCOA2", "BCL2L1", "TYK2", "BRAF", "EGFR", "LRP1B", "MSH3"),
           drawRowBar = TRUE, drawColBar = TRUE,
           clinicalFeatures = c("individualID", "tumorType", "sampleType"),
           sortByAnnotation = TRUE, annotationColor = color_list, groupAnnotationBySize = FALSE, GeneOrderSort = TRUE,
           removeNonMutated = FALSE, logColBar = FALSE,
           SampleNamefont = 5,annotationFontSize = 1.8, fontSize = 1.0,legendFontSize = 1.8,
           sepwd_genes = 1.5, sepwd_samples = 2.5,
           titleFontSize = 1.5, keepGeneOrder = TRUE, bgCol = "white", borderCol = "white")
dev.off()

oncoplot(jhu_public_maf_all, genes =c("NF1","TP53", "CDKN2A", "EED", "SUZ12", "EPC1", "ASTE1", "DAXX", "ATRX", "PCLO", "TYK2", "BRAF", "EGFR", "LRP1B", "MSH3"),
           drawRowBar = TRUE, drawColBar = TRUE,
           clinicalFeatures = c("individualID", "tumorType", "sampleType"),
           sortByAnnotation = TRUE, annotationColor = color_list, groupAnnotationBySize = FALSE, GeneOrderSort = TRUE,
           removeNonMutated = FALSE, logColBar = FALSE,
           SampleNamefont = 5,annotationFontSize = 1.8, fontSize = 1.0,legendFontSize = 1.8,
           titleFontSize = 1.5, keepGeneOrder = TRUE, bgCol = "white", borderCol = "white")
```

## Figure 2

Plot to show that cell line mutations match the sample. It'd be nice to see the NF1 and Pik3ca mutations for the 2-031 samples as a proof of cell line data. Our oncoplot for sample 2-031 does not capture any SNPs in NF1, possibly due to copy number variation in NF1 which is not reflected in VCFs. We are proceeding with CNV analysis for 2-031 for confirmation. 

We were able to crosscheck other samples in the bank with their clinical sequencing data (provided by Dr. Pratilas in https://app.box.com/file/530606854188), all the splice variants were confirmed but any large deletions were not captured in the VCF/MAF data files. This further confirms the need for CNV analysis for samples with these kind of deletions.

```{r sample 2-031, echo= FALSE, eval=TRUE, fig.height=8,fig.width=8, results='show', message=FALSE, warning=FALSE}

#Select specific sample files (select only the normal tissue and MPNST tissue and cell line samples)
sample_2031 <- jhu_public_clindata[(jhu_public_clindata$individualID == "2-031" & jhu_public_clindata$specimenID != "2-031 Xenograft"),]

#Make Maf file for sample 2-031
jhu_sample_2031 <- subsetMaf(jhu_public_maf_all, tsb = c(sample_2031$Tumor_Sample_Barcode)) 

# Make oncoplot to show the SNPs and mutation of the samples in various genes of interest
print("Fig 1 : Sample 2-031")
pdf("Sample_2-031.pdf", width=20, height = 8)
oncoplot(jhu_sample_2031, genes = c("NF1","PIK3CA", "PRC2","MEK1", "MEK2", "SUZ12", "ASTE1", "EGFR", "LRP1B", "MSH3", "PDCD1", "S1PR2", "EED", "EZH2"),
           drawRowBar = TRUE, drawColBar = TRUE,
           clinicalFeatures = c("tumorType", "sampleType","individualID"),
           sortByAnnotation = TRUE, annotationColor = color_list, groupAnnotationBySize = FALSE, 
           removeNonMutated = FALSE, logColBar = FALSE,
           SampleNamefont = 5,annotationFontSize = 1.5, fontSize = 1,legendFontSize = 1.8,
           titleFontSize = 1.0, keepGeneOrder = TRUE, GeneOrderSort = TRUE, bgCol = "white", borderCol = "white")
dev.off()

oncoplot(jhu_sample_2031, genes = c("NF1","PIK3CA", "PRC2","MEK1", "MEK2", "SUZ12", "ASTE1", "EGFR", "LRP1B", "MSH3", "PDCD1", "S1PR2", "EED", "EZH2"),
           drawRowBar = TRUE, drawColBar = TRUE,
           clinicalFeatures = c("tumorType", "sampleType","individualID"),
           sortByAnnotation = TRUE, annotationColor = color_list, groupAnnotationBySize = FALSE, 
           removeNonMutated = FALSE, logColBar = FALSE,
           SampleNamefont = 5,annotationFontSize = 1.5, fontSize = 1,legendFontSize = 1.8,
           titleFontSize = 1.0, keepGeneOrder = TRUE, GeneOrderSort = TRUE, bgCol = "white", borderCol = "white")
```
