---
title: "RNAseq_Analysis"
author: "Xengie Doan"
date: '`r Sys.Date()`'
output:
  html_document:
    code_folding: hide
    highlight: haddock
    latex_engine: xelatex
    theme: cerulean
    toc: yes
    toc_depth: 4
    toc_float: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, tidy.opts=list(width.cutoff=50), tidy=TRUE, fig.width=12, fig.height=8,fig.path = "/Users/xdoan/Shell/R/MPNSTAnalysis/analysis/figs/", dev=c('png', 'pdf'))

pkgs = c("DESeq2", "tidyverse", "RColorBrewer", "pheatmap", "ggplot2", "plotly", "biomaRt", "org.Hs.eg.db", "AnnotationDbi", "clusterProfiler", "reshape2", "knitr" ) #, "profvis") for profiling
lapply(pkgs, library, character.only = TRUE)
```

### load files
```{r}
# source("~/Shell/MPNSTAnalysis/data_access/RNASeqData.R")

load("data/salmon.result.metadata.Rda")
syn.query <- salmon.result[salmon.result$specimenID != "2-025 Neurofibroma",]
metadata.df <- syn.query[,colSums(is.na(syn.query))<nrow(syn.query)]
cols.keep <- c("specimenID","id","individualID","sex","tumorType","isCellLine","isPrimaryCell","tissue","experimentalCondition","study","study", "transplantationType")
metadata.df <- metadata.df[,cols.keep]

# load("data/txi.RData")
rownames(metadata.df) <- metadata.df$specimenID
```

###filter for protein coding genes
From the transcript counts step to make sure no unprocessed transcripts, etc are included in the counts. Also only includes genes with gene symbols. 
```{r}
# library(R.utils)
# path=synGet('syn18134565')$path
# R.utils::gunzip(path,overwrite=T)
# system(paste("grep protein_coding",gsub(".gz","",path),"|cut -d '|' -f 2 |uniq > gencode.v29.transcripts.txt"))

#now parse the headers
# genes=read.table('gencode.v29.transcripts.txt', colClasses = "character")
# save(genes, file ="protein_coding_ens_gene_id_ver.Rda")
# load("protein_coding_ens_gene_id_ver.Rda")
#ensembl gene id version

# txi$abundance <- txi$abundance[which(rownames(txi$abundance) %in% genes$V1),]
# txi$counts <- txi$counts[which(rownames(txi$counts) %in% genes$V1),]
# txi$length <- txi$length[which(rownames(txi$length) %in% genes$V1),]

### from sara's script https://github.com/sgosline/NEXUS/blob/master/analysis/2019-03-13/rnaSeqDiffEx.R
# expData <- synGet('syn18421359')
# expData <- read.csv(expData$path)
# expData <- expData %>% subset(Symbol != "" & study == "JHU NTAP Biobank" | study == "Preclinical NF1-MPNST Platform Development")
# expData<-expData%>%mutate(roundedCounts=round(totalCounts))

# save(expData, file = "data/prot_coding_counts_expData.Rda")
load("data/prot_coding_counts_expData.Rda")

data.mat=reshape2::acast(expData,Symbol~specimenID,value.var="roundedCounts")
missing=which(apply(data.mat,1,function(x) any(is.na(x))))
if(length(missing)>0)
  data.mat=data.mat[-missing,]

### manually change name of 2-004 specID
colnames(data.mat)[8] <- "2-004 Plexiform Neurofibroma"

### reorder data.mat
data.mat <- data.mat[,rownames(metadata.df)]
```

###filter for pNF and MPNST samples
Removed 2-009-Cell-Line and 2-009-Xenograft because of alleged mouse contamination
```{r}
### just pNF MPNST
### also remove 2-009-Cell-Line and 2-009-Xenograft bc mouse contamination
metadata.df_filtered <- metadata.df %>%
  filter(tumorType=="Plexiform Neurofibroma" | tumorType =="Malignant Peripheral Nerve Sheath Tumor") %>%
  filter(specimenID != "2-009 Cell Line" & specimenID != "2-009 Xenograft")

###shorten words and remove spaces for downstream
metadata.df_filtered$tumorType <- gsub("Plexiform Neurofibroma", "pNF", metadata.df_filtered$tumorType)
metadata.df_filtered$tumorType <- gsub("Malignant Peripheral Nerve Sheath Tumor", "MPNST", metadata.df_filtered$tumorType)

filtered_ids <- metadata.df_filtered$specimenID

data.mat <- data.mat[, filtered_ids]
# txi[[1]] <- txi[[1]][,filtered_ids]
# txi[[2]] <- txi[[2]][,filtered_ids]
# txi[[3]] <- txi[[3]][,filtered_ids]

## change metadata specimen IDs
metadata.df_filtered$specimenID <- gsub(" Plexiform Neurofibroma", "pNF", metadata.df_filtered$specimenID)
metadata.df_filtered$specimenID <- gsub(" Malignant Peripheral Nerve Sheath Tumor", "MPNST", metadata.df_filtered$specimenID)
metadata.df_filtered$specimenID <- gsub(" ", "-", metadata.df_filtered$specimenID)

### change colname in txi to match
# colnames(txi[[1]]) <- metadata.df_filtered$specimenID
# colnames(txi[[2]]) <- metadata.df_filtered$specimenID
# colnames(txi[[3]]) <- metadata.df_filtered$specimenID
colnames(data.mat) <- metadata.df_filtered$specimenID

rownames(metadata.df_filtered) <- metadata.df_filtered$specimenID
```


### run DESEQ
filter out counts < 10 and add study as a covariate to help with batch effects. set alpha to .05

```{r, message = FALSE, warning = FALSE}
# detach("package:synapser", unload=TRUE)
# unloadNamespace("PythonEmbedInR")
dds <- DESeqDataSetFromMatrix(countData = data.mat,
  colData = metadata.df_filtered,
  design= ~ study + tumorType)

# dds <- DESeqDataSetFromTximport(
#   txi = txi,
#   colData = metadata.df_filtered,
#   design = ~ study + tumorType) #batch effect of which study sequenced

###filter out reads with low counts
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

dds <- DESeq(dds)
# save(dds, file = "data/dds.Rda")
# load("data/dds.Rda")
res.1 <- results(dds, alpha = .1)
res.05 <- results(dds, alpha = .05)
# save(res, file = "data/res.Rda")
# load("data/res.Rda")
resultsNames(dds)
# resLFC <- lfcShrink(dds, coef = 3, type="apeglm") # plot looks weird with this type
resLFC <- lfcShrink(dds, coef = 3, type="normal")
```

Here the p-value is 0.1
```{r}
summary(res.1)
```
Here the p-value cutoff is 0.05
```{r}
summary(res.05)
```

### check outliers
```{r outlierCheck}
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds)[["cooks"]]), range=0, las=2)
```

### check dispersion
Dispersion is a metric for how much the variance differs from the mean in a negative binomial distribution. 
```{r dispersionCheck}
plotDispEsts(dds)
```

### MA plots
```{r MAplot}
plotMA(res.05, ylim=c(-2,2))
```

logFC shrink
```{r MAplotLFC}
plotMA(resLFC, ylim=c(-2,2))
```

```{r}
#already have gene names from Sara
res_df = as.data.frame(dplyr::mutate(as.data.frame(res.05), sig=ifelse(res.05$padj<0.05, "FDR<0.05", "Not Sig") ), row.names=rownames(res.05))

mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
# save(mart, file = "data/mart.Rda")
# load("data/mart.Rda")
# geneNames <- rownames(res_df)
# # listFilters(mart)
# 
# gene_list <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol"),values=geneNames,mart= mart)
# 
# res_df$ensembl_gene_id_version <- rownames(res_df)
# 
# res_df_geneName <- inner_join(res_df, gene_list, by = "ensembl_gene_id_version")
# res_df_geneName <- res_df_geneName[order(res_df_geneName$pvalue),]

### remove mitochondrial genes
# res_df_geneName <- res_df %>% mutate(hgnc_symbol = rownames(res_df))
# res_df_geneName <- res_df_geneName[ -c( grep("^MT-", res_df_geneName$hgnc_symbol) ),]

write.csv(res_df[1:1000,], file = "output/pNF_v_MPNST_geneList1000.csv", row.names = TRUE)
write.csv(res_df, file = "output/pNF_v_MPNST_geneList.csv", row.names = TRUE)
```

### heatmap of count matrix 
normalized by variance stabilizing transformation and displaying the top 50 highest expressed genes
```{r vsdHeatmap, fig.height=9, fig.width=9}
vsd <- vst(dds, blind=FALSE)
# save(vsd, file="data/vsd_remove2009contaminants_protCodingOnly.Rda")
#rename gene
vsd_matrix <- SummarizedExperiment::assay(vsd)

# select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing=TRUE)[1:50]
select <- order(rowMeans(vsd_matrix),decreasing=TRUE)[1:50]

top_genes <- vsd_matrix[select,]
# ## df to change rownames 
# vsd_df <- as.data.frame(vsd_matrix[select,])
# vsd_df$ensembl_gene_id_version <- rownames(vsd_df)
# vsd_df <- inner_join(vsd_df, gene_list, by = "ensembl_gene_id_version")
# 
# ## add NA to blanks
# vsd_df$hgnc_symbol <- gsub("^$", NA, vsd_df$hgnc_symbol)
# ## adds ens gene name to genes without hgnc symbol
# 
# rownames(vsd_df) <- coalesce(vsd_df$hgnc_symbol, vsd_df$ensembl_gene_id_version)
# 
# # remove gene name rows
# vsd_df <- vsd_df[,1:24]

# vsd_matrix <- as.matrix(vsd_df)

df <- as.data.frame(colData(dds)[,c("study", "tumorType", "sex", "transplantationType")])

pheatmap(top_genes, cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, annotation_col=df, color = rev(brewer.pal(11, "Spectral")))
```
We see a batch effect present because the samples group by study. We will visualize and correct for this downstream. 

### !to do 
```{r}
# plot top 20 gene expression
# look at list of marker/known genes
```


### heatmap of samp to samp distance
```{r heapmap_samp_to_samp}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$specimenID, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors, show_rownames = TRUE)
```



### pca by sample and tumorType
```{r pca_samp_tumor, fig.height=8, fig.width=8, warning = FALSE}
# load("data/vsd_remove2009contaminants_protCodingOnly.Rda")
p <- plotPCA(vsd, intgroup=c("individualID", "tumorType", "transplantationType", "isCellLine"), returnData= TRUE)
percentVar <- round(100 * attr(p, "percentVar"))
ggplot(p, aes(PC1, PC2, color = individualID, shape= tumorType, label = coalesce(vsd$transplantationType,ifelse(vsd$isCellLine == "true", "cellLine", NA)) )) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + ggrepel::geom_text_repel()
```

We are going to use removeBatchEffect from the limma package. 

#### removing study
```{r pca_noConsort, fig.height=8, fig.width=8}
load("data/vsd_remove2009contaminants_protCodingOnly.Rda")
mat <- assay(vsd)
#study is a big batch effect
mat <- limma::removeBatchEffect(mat, vsd$study)

assay(vsd) <- mat

p <- plotPCA(vsd, intgroup=c("individualID", "tumorType", "transplantationType", "isCellLine"), returnData= TRUE)
percentVar <- round(100 * attr(p, "percentVar"))
ggplot(p, aes(PC1, PC2, color = individualID, shape= tumorType, label = coalesce(vsd$transplantationType,ifelse(vsd$isCellLine == "true", "cellLine", NA)) )) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed() + ggrepel::geom_text_repel()
```

#### removing study and cell line
```{r pca_noConsortCellLine, fig.width=8, fig.height=8}
load("data/vsd_remove2009contaminants_protCodingOnly.Rda")
mat <- assay(vsd)
#study is a big batch effect
mat <- limma::removeBatchEffect(mat, vsd$study, vsd$isCellLine)
assay(vsd) <- mat

p <- plotPCA(vsd, intgroup=c("individualID", "tumorType", "transplantationType", "isCellLine"), returnData= TRUE)
percentVar <- round(100 * attr(p, "percentVar"))
ggplot(p, aes(PC1, PC2, color = individualID, shape= tumorType, label = coalesce(vsd$transplantationType,ifelse(vsd$isCellLine == "true", "cellLine", NA)) )) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed() + ggrepel::geom_text_repel()
```

#### removing xenograft status and cell line
```{r pca_noXenoCellLine, fig.width=8, fig.height=8}
load("data/vsd_remove2009contaminants_protCodingOnly.Rda")
mat <- assay(vsd)
vsd$transplantationType[is.na(vsd$transplantationType)] <- "no"

#study is a big batch effect
mat <- limma::removeBatchEffect(mat, vsd$transplantationType, vsd$isCellLine)
assay(vsd) <- mat

p <- plotPCA(vsd, intgroup=c("individualID", "tumorType", "transplantationType", "isCellLine"), returnData= TRUE)
percentVar <- round(100 * attr(p, "percentVar"))

#rename xeno no back to NA
vsd$transplantationType <- gsub("^no", NA, vsd$transplantationType)

ggplot(p, aes(PC1, PC2, color = individualID, shape= tumorType, label = coalesce(vsd$transplantationType,ifelse(vsd$isCellLine == "true", "cellLine", NA)) )) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  coord_fixed() + ggrepel::geom_text_repel()
ggplotly()
```

#### removing study, xenograft, and is cell line
```{r pca_noConsortXenoCellLine, fig.width=8, fig.height=8, warning = FALSE}
load("data/vsd_remove2009contaminants_protCodingOnly.Rda")
mat <- assay(vsd)

vsd$xenoCellLine <- paste(vsd$transplantationType, vsd$isCellLine, sep = ".")

#study is a big batch effect
mat <- limma::removeBatchEffect(mat, vsd$study, vsd$xenoCellLine)
assay(vsd) <- mat

p <- plotPCA(vsd, intgroup=c("individualID", "tumorType", "transplantationType", "isCellLine", "xenoCellLine"), returnData= TRUE)
percentVar <- round(100 * attr(p, "percentVar"))
ggplot(p, aes(PC1, PC2, color = individualID, shape= tumorType, label = coalesce(vsd$transplantationType,ifelse(vsd$isCellLine == "true", "cellLine", NA)) )) + geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + ggrepel::geom_text_repel()
```
BI386 is reported to be a post-chemotherapy sample so it doens't group with its xenograft counterpart. 

I'll use this corrected model because it has high % variance explained for eliminating most batch effects. For more detailed exploration please see [link](https://www.synapse.org/#!Synapse:syn18361348)


```{r}
### pretreatment or post treatment chemo 
```


### volcano plot
```{r volnano_plot}
###https://www.biostars.org/p/282295/
par(mar=c(5,5,5,5), cex=1.0, cex.main=1.4, cex.axis=1.4, cex.lab=1.4)

topT <- as.data.frame(res.05)

#Adjusted P values (FDR Q values)
with(topT, plot(log2FoldChange,  -log10(padj), pch=20, main="Volcano plot", cex=1.0, xlab=bquote(~Log[2]~fold~change), ylab=bquote(~-log[10]~p~adj)))
with(subset(topT, padj<0.05 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(padj), pch=20, col="red", cex=0.5))

#with(subset(topT, padj<0.05 & abs(log2FoldChange)>2), text(log2FoldChange, -log10(padj), labels=subset(rownames(topT), topT$padj<0.05 & abs(topT$log2FoldChange)>2), cex=0.8, pos=3))

#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-2, col="black", lty=4, lwd=2.0)
abline(v=2, col="black", lty=4, lwd=2.0)
abline(h=-log10(max(topT$pvalue[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
```
```{r}
###https://shiring.github.io/rna-seq/deseq2/teaching/2016/09/29/DESeq2-course
# res_df = as.data.frame(dplyr::mutate(as.data.frame(res), sig=ifelse(res$padj<0.05, "FDR<0.05", "Not Sig") ), row.names=rownames(res))
# 
# DEgenes_DESeq <- res_df[which(abs(res_df$log2FoldChange) > log2(1.5) & res$padj < 0.05),]
# 
# p <- ggplot2::ggplot(res_df[order(res_df$log2FoldChange),], ggplot2::aes(log2FoldChange, -log10(pvalue))) +
#   ggplot2::geom_point(ggplot2::aes(col = sig)) +
#   ggplot2::scale_color_manual(values = c("red", "black")) +
#   ggplot2::ggtitle("Volcano Plot of DESeq2 analysis")
# p
# p + ggrepel::geom_text_repel(data=res_df[order(res_df$log2FoldChange),][1:10, ], ggplot2::aes(label=rownames(res_df[order(res_df$log2FoldChange),][1:10, ])))
```

###Heatmap of batch corrected genes


```{r, results= FALSE}
# I used the rlog transformation, transforming the counda data to the log2 scale.The regularization is on the log fold changes of the count for each sample over an intercept, for each gene. As nearby count values for low counts genes are almost as likely as the observed count, the rlog shrinkage is greater for low counts. For high counts, the rlog shrinkage has a much weaker effect. This might be why we see this pattern where it is only high log2FC genes int the top 50. 
# rld <- rlog(dds, blind=F)
# save(rld, file ="data/rld.Rda")
load("data/rld.Rda")

### regularized log transformation 
exp_matrix <- SummarizedExperiment::assay(rld)
### top 100
select <- order(rowMeans(exp_matrix), decreasing=TRUE)[1:20]

## df to change rownames 
top100 <- exp_matrix[select,]
# top100$ensembl_gene_id_version <- rownames(top100)
# top100 <- inner_join(top100, gene_list, by = "ensembl_gene_id_version")
# 
# ## add NA to blanks
# top100$hgnc_symbol <- gsub("^$", NA, top100$hgnc_symbol)
# ## adds ens gene name to genes without hgnc symbol
# 
# rownames(top100) <- coalesce(top100$hgnc_symbol, top100$ensembl_gene_id_version)

# remove gene name rows 
# top100 <- top100[,1:24]

# top100_matrix <- as.matrix(top100)

annotation_data <- as.data.frame(colData(rld)["tumorType"] )
# pheatmap(exp_matrix[select,], annotation_col = annotation_data)
# colors <- grDevices::colorRampPalette( rev(RColorBrewer::brewer.pal(11, "Spectral")) )(255)
# pheatmap(top100_matrix, annotation_col = annotation_data, col = colors)
```

```{r, fig.height=8, fig.width=8, eval = FALSE}
# num_breaks <- 30
# myColor <- colorRampPalette(c("yellow", "white","blue"))(num_breaks)
# 
# myBreaks <- c(seq(min(top100_matrix, na.rm=T), 0, length.out=ceiling(num_breaks/3) + 1),
#               seq((-1*min(top100_matrix, na.rm=T))/num_breaks, (-1*min(top100_matrix, na.rm=T)), length.out=floor(num_breaks/3)),
#               seq((-1*min(top100_matrix, na.rm=T))+max(top100_matrix, na.rm=T)/num_breaks, max(top100_matrix, na.rm=T), length.out=floor(num_breaks/3))
#               )
# 
# myBreaks2 <- c(seq(min(top100_matrix), 0, length.out=ceiling(num_breaks/2) + 1),
#               seq(max(top100_matrix)/num_breaks, max(top100_matrix), length.out=floor(num_breaks/2)))
# annotation_data <- as.data.frame(colData(rld)["study"] )
# pheatmap(top100_matrix ,color=myColor, breaks=myBreaks2, annotation_col = annotation_data)
# 
# df$cancer <- as.factor(rep("esophageal", nrow(df)))
```


```{r}
#from http://folk.uio.no/jonbra/MBV-INF4410_2017/exercises/2017-12-07_R_DESeq2_exercises_without_results.html
# mat <- assay(rld)
# mat_df <- as.data.frame(mat)
# ## get protein coding genes
# mat_prot_df <- subset(mat_df, rownames(mat_df) %in% genes$V1)
# 
# mat_prot_df <- mat_prot_df[ head(order(res.05[rownames(res.05) %in% genes$V1,]$padj),30), ] # select the top 30 genes with the lowest padj
# mat_prot_df = mat_prot_df - rowMeans(mat_prot_df) # Subtract the row means from each value to standardize
# 
# ## df to change rownames 
# mat_prot_df$ensembl_gene_id_version <- rownames(mat_prot_df)
# 
# mat_prot_df <- inner_join(mat_prot_df, gene_list, by = "ensembl_gene_id_version")
# 
# ## add NA to blanks
# mat_prot_df$hgnc_symbol <- gsub("^$", NA, mat_prot_df$hgnc_symbol)
# ## adds ens gene name to genes without hgnc symbol
# rownames(mat_prot_df) <- coalesce(mat_prot_df$hgnc_symbol, mat_prot_df$ensembl_gene_id_version)
# 
# # remove gene name rows
# mat_prot_matrix <- as.matrix(mat_prot_df[,1:24])
# 
# df = as.data.frame(colData(rld)[,c("study")]) # Create a dataframe with a column of the conditions
# colnames(df) = "study" # Rename the column header
# rownames(df) = colnames(mat) # add rownames
# # and plot the actual heatmap
# pheatmap(mat_prot_matrix, annotation_col=df)
```

Here are the top 50 genes with the lowest p adjusted value
```{r heatmap_top50_batchCorr,fig.height=9, fig.width=9}
mat <- assay(vsd)
mat_df <- as.data.frame(mat)
## get protein coding genes
# mat_prot_df <- subset(mat_df, rownames(mat_df) %in% genes$V1)

mat_prot_df <- mat_df[ head(order(res.05[rownames(res.05),]$padj),50), ] # select the top 50 genes with the lowest padj
mat_prot_df = mat_prot_df - rowMeans(mat_prot_df) # Subtract the row means from each value to standardize

## df to change rownames 
# mat_prot_df$ensembl_gene_id_version <- rownames(mat_prot_df)
# 
# mat_prot_df <- inner_join(mat_prot_df, gene_list, by = "ensembl_gene_id_version")

## add NA to blanks
# mat_prot_df$hgnc_symbol <- gsub("^$", NA, mat_prot_df$hgnc_symbol)
## adds ens gene name to genes without hgnc symbol
# rownames(mat_prot_df) <- coalesce(mat_prot_df$hgnc_symbol, mat_prot_df$ensembl_gene_id_version)

# remove gene name rows
mat_prot_matrix <- as.matrix(mat_prot_df)

df = as.data.frame(colData(rld)[,c("study", "sex", "tumorType", "isCellLine", "transplantationType")]) # Create a dataframe with a column of the conditions
rownames(df) = colnames(mat) # add rownames
# and plot the actual heatmap
pheatmap(mat_prot_matrix, annotation_col=df, color = rev(brewer.pal(11, "Spectral")))

### add sex, cell line, tumortype, xenograft, 
### heatmap based on top GO terms
```


### GO Terms

```{r}
#https://shiring.github.io/rna-seq/deseq2/teaching/2016/09/29/DESeq2-course
#get entrezid
entrez_list <- getBM(filters= "hgnc_symbol", attributes= c("hgnc_symbol", "entrezgene"),values=rownames(mat_df),mart= mart)
mat_df$hgnc_symbol <- rownames(mat_df)
mat_df <- inner_join(mat_df, entrez_list, by = "hgnc_symbol")
```

This is the GO profile using a biological processes subontology
```{r go_profile, warning = FALSE}
OrgDb <- org.Hs.eg.db # can also be other organisms
gene <- na.omit(mat_df$entrezgene)

# Group GO
ggo <- clusterProfiler::groupGO(gene     = as.character(gene),
                                OrgDb    = OrgDb,
                                ont      = "BP",
                                level    = 3,
                                readable = TRUE)

barplot(ggo, drop=TRUE, showCategory=12)
```
Here are the top 12  profiles. 

### GO over-representation test
Using a p value and q value of .05, this returns the enrichment GO categories. Shown are the top 6
```{r}
# GO over-representation test
ego <- clusterProfiler::enrichGO(gene          = gene,
                                 OrgDb         = OrgDb,
                                 ont           = "BP",
                                 pAdjustMethod = "BH",
                                 pvalueCutoff  = 0.05,
                                 qvalueCutoff  = 0.05, 
                                 readable      = TRUE)

DT::datatable(head(summary(ego)[,-8]), rownames = FALSE)
```
```{r, warning = FALSE}
top6id <- head(as.data.frame(ego)[,-8])$ID

write.csv(ego[1:1000,],file="output/pNFvMPNST_GO_overRep1000.csv")
write.csv(ego,file="output/pNFvMPNST_GO_overRep.csv", row.names = FALSE)
```


```{r}
go_entrez <- as.numeric(unlist(attributes(ego)$geneSets[ top6id]))

gene_go <- subset(mat_df, mat_df$entrezgene %in% go_entrez)
gene_go <- unique(gene_go)
### get entrez id of dds and subset by that and put into heatmap
```

### Heatmap of genes in top6 GO overrepresented sets
top 50 genes by padj value that are in the GO overrepresented sets that are standardized by subtracting the rowmean from the count. 
```{r heatmap_goOverRep}
### for original vsd
mat <- assay(vsd)
mat_df <- as.data.frame(mat)
mat_df$hgnc_symbol <- rownames(mat_df)
mat_df <- inner_join(mat_df, entrez_list, by = "hgnc_symbol")

gene_go <- subset(mat_df, mat_df$entrezgene %in% go_entrez)
gene_go <- unique(gene_go)
###

select <- order(res.05[rownames(res.05),]$padj)[1:500]
select <- rownames(res.05[select,])

# select <- order(rowMeans(gene_go[1:24]),decreasing=TRUE)[1:100]
top_gene_go <- gene_go[ gene_go$hgnc_symbol %in% select,][1:50,]
top_gene_go[,1:24] = top_gene_go[,1:24] - rowMeans(top_gene_go[,1:24]) # Subtract the row means from each value to standardize

rownames(top_gene_go) <- top_gene_go$ensembl_gene_id_version

# top_gene_go <- inner_join(top_gene_go, gene_list, by = "ensembl_gene_id_version")

## add NA to blanks
# top_gene_go$hgnc_symbol <- gsub("^$", NA, top_gene_go$hgnc_symbol)
# ## adds ens gene name to genes without hgnc symbol
# rownames(top_gene_go) <- coalesce(top_gene_go$hgnc_symbol, top_gene_go$ensembl_gene_id_version)

# remove gene name rows
top_gene_go <- as.matrix(top_gene_go[,1:24])

# df = as.data.frame( colData(rld)[,c("study")] ) # Create a dataframe with a column of the conditions

annotation_col <- data.frame(
        batch = colData(rld)[,c("study")],
        tumorType = colData(rld)[,c("tumorType")],
        sex = colData(rld)["sex"], 
        transplantationType = colData(rld)["transplantationType"], 
        isCellLine = colData(rld)["isCellLine"])

rownames(annotation_col) = colnames(mat) # add rownames
# and plot the actual heatmap
pheatmap(top_gene_go, annotation_col=annotation_col, show_colnames = F, color = rev(brewer.pal(11, "Spectral")))
```

###Plot of GO enriched genes by category
Here is a barplot of the GO enriched genes by category
```{r barplot_GOenrich, fig.width=10}
barplot(ego, showCategory=25)
```

```{r dotplot_GOenrich, fig.height=10, fig.width=10}
clusterProfiler::dotplot(ego, showCategory=25)
```
Above is a dotplot of the GO enriched genes by category and showing the gene ratio and padj. 

###KEGG pathways
```{r, warning=FALSE}
## KEGG
kk <- clusterProfiler::enrichKEGG(gene         = gene,
                 organism     = 'hsa',
                 pAdjustMethod = "BH",
                 pvalueCutoff = 0.05,
                 qvalueCutoff  = 0.05)
DT::datatable(head(summary(kk)[,-8]), rownames = FALSE)
```
```{r}
write.csv(kk,file="output/pNFvMPNST_KEGG.csv", row.names = FALSE)
### cluster on GSEA pathways
```

```{r}
# methods paper for biobank data
# triplicate data, 5 patients, whats the stepwise change from rna/exome
```


